src/tactic/ring_exp.lean

```diff
@@ -0,0 +1,1430 @@
+
+  Subtraction cancels out identical terms, but division does not.
+  That is: `a - a = 0 := by ring_exp` solves the goal,
+  but `a / a := 1 by ring_exp` doesn't.
+  Note that `0 / 0` is generally defined to be `0`,
+  so division cancelling out is not true in general.
+
+  Multiplication of powers can be simplified a little bit further:
+  `2 ^ n * 2 ^ n = 4 ^ n := by ring_exp` could be implemented
+  in a similar way that `2 * a + 2 * a = 4 * a := by ring_exp` already works.
+  This feature wasn't needed yet, so it's not implemented yet.
+
+  ## Tags
+
+  ring, semiring, exponent, power
+-/
+
+-- The base ring `α` will have a universe level `u`.
+-- We do not introduce `α` as a variable yet,
+-- in order to make it explicit or implicit as required.
+universes u
+
+namespace tactic.ring_exp
+open nat
+
+/--
+  The `atom` structure is used to represent atomic expressions:
+  those which `ring_exp` cannot parse any further.
+
+  For instance, `a + (a % b)` has `a` and `(a % b)` as atoms.
+  The `ring_exp_eq` tactic does not normalize the subexpressions in atoms,
+  but `ring_exp` does if `ring_exp_eq` was not sufficient.
+
+  Atoms in fact represent equivalence classes of expressions,
+  modulo definitional equality.
+  The field `index : ℕ` should be a unique number for each class,
+  while `value : expr` contains a representative of this class.
+  The function `resolve_atom` determines the appropriate atom
+  for a given expression.
+-/
+meta structure atom : Type := (value : expr) (index : ℕ)
+namespace atom
+/--
+  The `eq` operation on `atom`s works modulo definitional equality,
+  ignoring their `value`s.
+  The invariants on `atom` ensure indices are unique per value.
+  Thus, `eq` indicates equality as long as the `atom`s come from the same context.
+-/
+meta def eq (a b : atom) : bool := a.index = b.index
+/--
+  We order `atom`s on the order of appearance in the main expression.
+-/
+meta def lt (a b : atom) : bool := a.index < b.index
+meta instance : has_repr atom := ⟨λ x, "(atom " ++ repr x.2 ++ ")"⟩
+end atom
+
+section expression
+/-!
+  ## `expression` section
+  In this section, we define the `ex` type and its basic operations.
+
+  First, we introduce the supporting types `coeff`, `ex_type` and `ex_info`.
+  For understanding the code, it's easier to check out `ex` itself first,
+  then refer back to the supporting types.
+
+  The arithmetic operations on `ex` need additional definitions,
+  so they are defined in a later section.
+-/
+
+/--
+  Coefficients in the expression are stored in a wrapper structure,
+  allowing for easier modification of the data structures.
+  The modifications might be caching of the result of `expr.of_rat`,
+  or using a different meta representation of numerals.
+-/
+@[derive decidable_eq]
+structure coeff : Type := (value {} : ℚ)
+
+/--
+  The `ex` type is structured according to a parameter `et : ex_type`.
+
+  Instead of allowing each operand to have arbitrary subexpressions as argument,
```

Comment:
I'm a little confused by this comment. It seems to be describing `ex`, not `ex_type`, no?